internal/quickfort/query.lua

```diff
@@ -5,12 +5,117 @@ if not dfhack_flags.module then
     qerror('this script cannot be called directly')
 end
 
+local gui = require('gui')
+local guidm = require('gui.dwarfmode')
 local quickfort_common = reqscript('internal/quickfort/common')
 local log = quickfort_common.log
+local quickfort_aliases = reqscript('internal/quickfort/aliases')
+local quickfort_keycodes = reqscript('internal/quickfort/keycodes')
+
+local common_aliases_filename = 'hack/data/quickfort/aliases-common.txt'
+local user_aliases_filename = 'dfhack-config/quickfort/aliases.txt'
+
+local function load_aliases()
+    -- ensure we're starting from a clean alias stack, even if the previous
+    -- invocation of this function returned early with an error
+    quickfort_aliases.reset_aliases()
+    quickfort_aliases.push_aliases_csv_file(common_aliases_filename)
+    quickfort_aliases.push_aliases_csv_file(user_aliases_filename)
+end
+
+local function is_queryable_tile(pos)
+    local flags, occupancy = dfhack.maps.getTileFlags(pos)
+    return not flags.hidden and occupancy.building ~= 0
+end
+
+local function move_cursor(screen, overlay, pos)
+    overlay:moveCursorTo(pos)
+    -- wiggle the cursor so the building under the cursor gets properly selected
+    -- is there a better way to do this? I've tried:
+    -- - DwarfOverlay:simulateCursorMovement
+    -- - DwarfOverlay:selectBuilding
+    -- - setting the ui.main.mode to something else then back to QueryBuilding
+    if pos.y > 0 then
+        gui.simulateInput(screen, 'CURSOR_UP')
+        gui.simulateInput(screen, 'CURSOR_DOWN')
+    else
+        gui.simulateInput(screen, 'CURSOR_DOWN')
+        gui.simulateInput(screen, 'CURSOR_UP')
+    end
+end
+
+local function handle_modifiers(token, modifiers)
+    local token_lower = token:lower()
+    if token_lower == '{shift}' or
+            token_lower == '{ctrl}' or
+            token_lower == '{alt}' then
+        modifiers[token_lower] = true
+        return true
+    end
+    if token_lower == '{wait}' then
+        print('{wait} not yet implemented')
```

Comment:
Out of curiosity, what does this do? If it's a delay to allow DF to catch up, hopefully that won't be necessary with the script.

Reply:
Evidently I didn't read your comment very thoroughly before looking at the code. See below for ideas about what you could do for sleeping.

Reply:
I'm actually not sure what the original intent was in python quickfort. It could be that -- allowing the UI to catch up, and I agree that is completely unnecessary now. The other, potentially still useful function is to allow a user to see what the current screen looks like at that point in the keystroke playback for debugging reasons. I don't know if anyone actually uses this, but I'm aiming for python quickfort compatibility so unless there's a reason *not* to do it, I plan to support it.

I'm thinking there is a greater opportunity to implement a "debug mode" that allows a user to set breakpoints, poke around the UI, then resume (or halt) playback. but that's for a future PR : )